collect_app/src/main/java/org/odk/collect/android/formentry/FormSessionRepository.kt

```diff
@@ -37,8 +39,16 @@ class AppStateFormSessionRepository(application: Application) : FormSessionRepos
         getLiveData(id).value = FormSession(formController, form, instance)
     }
 
+    /**
+     * Ensure the object gets completely removed. Simply nullifying it might cause memory leaks.
+     * See: https://github.com/getodk/collect/issues/5777
+     */
     override fun clear(id: String) {
-        getLiveData(id).value = null
+        appState.clear(getKey(id))
```

Comment:
I don't have a lot of experience with garbage collection but this seems like a bug to me... with the old way of removing objects (by nullifying them), the session object (the main object) was garbage collected but not the properties inside eg. not the `formController` which in this case is the object that consumes a lot of memory. But if we remove the object from the map at all, then everything is fine.

Reply:
I guess the big difference here will be that setting the old version `getLiveData(id).value = null` would set the `LiveData` in the map to `null` which would propagate the value to any observers, but the `LiveData` itself wouldn't be garbage collected. The new code will remove the `LiveData` from the map meaning that it can be garbage collected.

My hunch would be that there is an observer that wasn't getting unregistered somehow. That would mean that it would be referenced by the `LiveData` and so if the `LiveData` wasn't garbage collected it wouldn't be. It'd be good to find the offender here so we can make sure we've fixed any other potential memory leaks from an observer hanging around.

Reply:
Ah! Looks like it could be `FormSaveViewModel`: https://github.com/getodk/collect/blob/3d66ab17f4d2724326fda3c94e251394bb1735bd/collect_app/src/main/java/org/odk/collect/android/formentry/saving/FormSaveViewModel.java#L106

We never unregister that observation, so the `FormSaveViewModel` will continue to be referenced by the `LiveData`. We do the correct clean up in `BackgroundAudioViewModel` (unregistering the observation in `onCleared`). I wonder if you can fix that and the memory leak will go away?

Reply:
`VieModel` should be destroyed after closing the activity so isn't it cleared automatically? I will test it but even if it's true it's very error-prone because it's easy to forget about clearing livedata...
I think in this case probably we should first call:
`getLiveData(id).value = null`
so that as you said null value is propagated and then
`appState.clear(getKey(id))`
to avoid memory leaks

Reply:
AppState (in this case) stores a map of `MutableLiveData<FormSession>` for given key. With the old way of clearing it we assigned null to that key. Shouldn't that be enough for GC to clear everything? If we nullify the livedata the observers should be gone too. The only change I implemented is removing livedata from the map instead of nullifying it. I'm not saying that your theory is wrong but still is that behavior ok because it still seems to me like a bug.

Reply:
I think from GC perspective I don't really get the difference between removing livedata from the map by setting the value to null vs removing it from the map completely. In both cases, we do not nullify the livedata directly.

Reply:
> Maybe it's because of the fact that getLiveData(id).value = null does not nullify the livedata directly but assigns null to the key that the livedata was represented by in the map... but still, such a livedata is then not used anywhere (apart from the ViewModel which should be destroyed after closing its activity) then so GC should be able to get rid of it?

With `getLiveData(id).value = null` we still end up with the map referencing the `LiveData` instance, it's just that the **value** of that `LiveData` is `null`. Given `AppState` is referenced by the `Application` instance (an instance of `Collect`), that `LiveData` will not be garbage collected. If there's an observer registered to the `LiveData`, we'd end up with a reference chain like this:


```
Collect -> AppState -> LiveData -> Observer
```


If the `Observer` is a `ViewModel`, you're right that it'll get destroyed when the Activity is destroyed, but only in the sense that the Activity will call `onCleared` and stop referencing it - the `ViewModel` would still be referenced by the `LiveData` which means that it wouldn't get garbage collected. It's Android "destruction" vs actual destruction in memory.

My hunch is that when you changed the line to `appState.clear(getKey(id))` you changed it so that the chain to `ViewModel` after Activity destruction would just be:

```
LiveData -> ViewModel
```

As nothing references the `LiveData` anymore, we have an "orphaned" graph that can be garbage collected. 

I could definitely be wrong about all this, but I think it's worth experimenting with an `onCleared` in `FormSaveViewModel` that unregisters it as an observer and seeing if that would also fix the memory leak. 

The current fix changes the behaviour of `FormStateRepository` subtly as `clear` now means that any subsequent calls to `get`/`set` with the same ID would deal with a different `LiveData` instance (meaning observations before the `clear` call wouldn't get updates).

Reply:
Ah right after calling `getLiveData(id).value = null` it's still livedata but with a null value, I assumed that it nullifies the whole livedata in the map ok so it's clear...
So probably it can't be garbage collected because despite the fact that the livedata has a null value it has an observer in the view model and that view model keeps a reference to formControler and it can't be removed.
It's very easy to make such mistakes.